plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -1521,4 +1458,27 @@ public int hashCode()
             return Objects.hash(baseColumnIdentity, path);
         }
     }
+
+    private record MergeRowIdBlockFactory(VariableWidthBlock filePath, IntArrayBlock partitionSpecId, VariableWidthBlock partitionData)
+            implements Function<Block, RowBlock>
+    {
+        private static Function<Block, RowBlock> create(Slice filePath, int partitionSpecId, Slice partitionData)
+        {
+            return new MergeRowIdBlockFactory(
+                    new VariableWidthBlock(1, filePath, new int[] {0, filePath.length()}, Optional.empty()),
+                    new IntArrayBlock(1, Optional.empty(), new int[] {partitionSpecId}),
+                    new VariableWidthBlock(1, partitionData, new int[] {0, partitionData.length()}, Optional.empty()));
+        }
+
+        @Override
+        public RowBlock apply(Block rowPosition)
+        {
+            return RowBlock.fromFieldBlocks(rowPosition.getPositionCount(), new Block[] {
+                    RunLengthEncodedBlock.create(filePath, rowPosition.getPositionCount()),
```

Comment:
Pre-existing

I am wondering whether the components `(_file, _pos, partition_spec_id, partition_data)` can be actually considered the row identifier.


With [add_files](https://trino.io/docs/current/connector/iceberg.html#add-files) we can add exactly the same file twice in an Iceberg table.

When doing the sequence of operations:
- add_file('myfile')
- add equality delete affecting the file 'myfile'
- add_file('myfile')

The second addition of 'myfile' is affected by the previous equality delete applied on 'myfile'

cc @ebyhr

Reply:
It's a known limitation of add_files procedure #23678.

Reply:
https://github.com/trinodb/trino/pull/24188 is PR to deny duplicate files in the procedure.